src/libprojectM/Renderer/MilkdropWaveform.cpp

```diff
@@ -4,393 +4,613 @@
+            }
+            break;
+        }
 
-		case DoubleLine://dual waveforms
-		{
+        case MilkdropWaveformMode::Line:
+        case MilkdropWaveformMode::DoubleLine:
+        case MilkdropWaveformMode::SpectrumLine: // Unfinished
+        {
+            if (mode == MilkdropWaveformMode::SpectrumLine)
+            {
+                samples = 256;
+            }
+            else
+            {
+                samples = NumWaveformSamples / 2;
+
+                if (samples > context.viewportSizeX / 3)
+                {
+                    samples /= 3;
+                }
+            }
+
+            wave1Vertices = new WaveformVertex[samples]();
+            if (mode == MilkdropWaveformMode::DoubleLine)
+            {
+                wave2Vertices = new WaveformVertex[samples]();
+            }
+
+            int sampleOffset = (NumWaveformSamples - samples) / 2;
+
+            float angle = PI * 0.5f * mysteryWaveParam; // from -PI/2 to PI/2
+            float dx = cosf(angle);
+            float dy = sinf(angle);
+
+            float edgeX[2]{
+                x * cosf(angle + PI * 0.5f) - dx * 3.0f,
+                x * cosf(angle + PI * 0.5f) + dx * 3.0f
+            };
+
+            float edgeY[2]{
+                x * sinf(angle + PI * 0.5f) - dy * 3.0f,
+                x * sinf(angle + PI * 0.5f) + dy * 3.0f
+            };
+
+            for (int i = 0; i < 2; i++)
+            {
+                for (int j = 0; j < 4; j++)
+                {
+                    float t;
+                    bool clip{ false };
+
+                    switch (j)
+                    {
+                        case 0:
+                            if (edgeX[i] > 1.1f)
+                            {
+                                t = (1.1f - edgeX[1 - i]) / (edgeX[i] - edgeX[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        case 1:
+                            if (edgeX[i] < -1.1f)
+                            {
+                                t = (-1.1f - edgeX[1 - i]) / (edgeX[i] - edgeX[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        case 2:
+                            if (edgeY[i] > 1.1f)
+                            {
+                                t = (1.1f - edgeY[1 - i]) / (edgeY[i] - edgeY[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        case 3:
+                            if (edgeY[i] < -1.1f)
+                            {
+                                t = (-1.1f - edgeY[1 - i]) / (edgeY[i] - edgeY[1 - i]);
+                                clip = true;
+                            }
+                            break;
+
+                        default:
+                            break;
```

Comment:
idk shouldn't we blow up here?

Reply:
Well, no. But remove the default, as it's not needed. Compiler will most probably unroll the loop in optimized builds.